Stripe webhook secret implimentation

TechNoNerd87 opened the pull request:
Summary

This PR adds optional Stripe Webhook signature verification to ensure improved security and full compatibility with existing Invoice Ninja installations.

Background

Stripe recommends verifying the Webhook Signing Secret for all webhook events.
However, many existing Invoice Ninja installations do not have this configured, and enabling verification without backward compatibility would immediately break webhook processing.

What This PR Does

Adds Stripe webhook signature verification

Fully backward compatible with all current systems

Verification occurs only if a Webhook Secret is defined in the admin portal

If the field is left blank, the webhook will be processed without verification (current behavior preserved)

Implementation Logic

Admin UI allows the Webhook Secret to remain empty

When empty → skip verification

When populated → verify using Stripe’s recommended signature method

Why This is Needed

Stripe explicitly recommends validating webhook signatures for security and fraud prevention.
This implementation brings Invoice Ninja closer to Stripe best practice while ensuring existing installations do not break.

Important Note for Administrators

When adding a new Stripe payment gateway, you must:

Leave the “Webhook Secret” field blank when creating the gateway

Save and retrieve the generated webhook URL

Add this URL in the Stripe Dashboard and generate the Webhook Secret

Return to Invoice Ninja and paste the Webhook Secret into the gateway settings

This is required because Stripe does not provide the webhook secret until after their endpoint is created.

TechNoNerd87 committed:
Add webhook secret and signature verification

Added webhook_secret property and implemented signature verification for Stripe webhook requests.

Signed-off-by: TechNoNerd87 <113461509+TechNoNerd87@users.noreply.github.com>

TechNoNerd87 committed:
Modify Stripe fields in PaymentLibrariesSeeder

Updated Stripe payment provider fields to include webhookSecret.

Signed-off-by: TechNoNerd87 <113461509+TechNoNerd87@users.noreply.github.com>

TechNoNerd87 committed:
Initialize webhook secret in processWebhookRequest

Add initialization for webhook secret handling in Stripe payment driver.

Signed-off-by: TechNoNerd87 <113461509+TechNoNerd87@users.noreply.github.com>

turbo124 commented:
Hi There,

Thanks for this. We would need to also include a migration to enable gateways that have already been created to have the webhookSecret prop applied to the config model.

TechNoNerd87 committed:
Refactor Stripe webhook request processing

Refactor webhook processing to simplify signature verification and log request data.

Signed-off-by: TechNoNerd87 <113461509+TechNoNerd87@users.noreply.github.com>

TechNoNerd87 committed:
Fix Stripe webhook signature verification handling

Signed-off-by: TechNoNerd87 <113461509+TechNoNerd87@users.noreply.github.com>

TechNoNerd87 committed:
Refactor StripePaymentDriver webhook handling

Removed assignment of webhook_secret and fixed syntax error in webhook event construction.

Signed-off-by: TechNoNerd87 <113461509+TechNoNerd87@users.noreply.github.com>

TechNoNerd87 committed:
Fix webhook signature verification logic

Signed-off-by: TechNoNerd87 <113461509+TechNoNerd87@users.noreply.github.com>

TechNoNerd87 committed:
Fix webhook signature verification handling

Signed-off-by: TechNoNerd87 <113461509+TechNoNerd87@users.noreply.github.com>

TechNoNerd87 commented:
I've updated and tested the code accordingly.

The implementation keeps the Stripe webhook secret optional, so existing installations remain fully compatible. If a gateway doesn't have a webhookSecret set, the webhook signature check is simply skipped. If the secret is populated, validation is enforced as expected.

This means current installs won’t break after updating. Administrators who want to enable verification can simply edit their Stripe gateway in the web UI and fill in the webhook secret field.

After applying the changes, I also needed to run:
php artisan db:seed --class=PaymentLibrariesSeeder

I did not need to run any migration or modify the database since the Stripe Webhook secret property already exists.

turbo124 commented:
@TechNoNerd87 it exisitng when creating a new stripe gateway, but users who have already configured a stripe gateway would not be able to access this field. I'll review now.

turbo124 commented on app/PaymentDrivers/StripePaymentDriver.php:
```diff
@@ -85,6 +85,8 @@ class StripePaymentDriver extends BaseDriver implements SupportsHeadlessInterfac
     public $stripe_connect_auth = [];
 
+    public $webhook_secret = "";
+
```
no need to create a class var

turbo124 commented on app/PaymentDrivers/StripePaymentDriver.php:
```diff
@@ -122,6 +124,8 @@ public function init()
             }
         } else {
+            $this->webhook_secret = $this->company_gateway->getConfigField('webhookSecret');
+            
```
no need for this, redundant

turbo124 reviewed (changes requested):
remove redundant class props.

TechNoNerd87 committed:
Remove webhook_secret from StripePaymentDriver

Removed webhook_secret property and its assignment.

Signed-off-by: TechNoNerd87 <113461509+TechNoNerd87@users.noreply.github.com>

turbo124 reviewed (approved):
Looks good. Thanks!
